machine/machine_core/ssh_connection.py

```diff
@@ -248,26 +255,28 @@ def _check_ssh_master(self):
                 return True
         return False
 
-    def _ensure_ssh_master(self):
+    def _ensure_ssh_master(self) -> None:
         if not self._check_ssh_master():
             self._start_ssh_master()
 
-    def __ssh_direct_opt_var(self, direct=False):
+    def __ssh_direct_opt_var(self, direct: bool = False) -> bool:
         return bool(os.getenv("TEST_SSH_DIRECT", direct))
 
-    def __execution_opts(self, direct=False):
+    def __execution_opts(self, direct: bool = False) -> List[str]:
         direct = self.__ssh_direct_opt_var(direct=direct)
         if direct:
             return ["-i", self.identity_file]
-        else:
+        elif self.ssh_master:
             return ["-o", "ControlPath=" + self.ssh_master]
+        return []
 
-    def execute(self, command, input=None, environment={},
-                stdout=subprocess.PIPE, quiet=False, direct=False, timeout=120,
-                ssh_env=["env", "-u", "LANGUAGE", "LC_ALL=C"], check=True):
+    def execute(self, command: Union[List[str], str], input: Optional[str] = None, environment: Dict[str, str] = {},
+                stdout: int = subprocess.PIPE, quiet: bool = False, direct: bool = False,
+                timeout: int = 120, ssh_env: List[str] = ["env", "-u", "LANGUAGE", "LC_ALL=C"],
+                check: bool = True) -> Optional[str]:
         """Execute a shell command in the test machine and return its output.
 
-            command: The string or argument list to execute by /bin/sh (still with shell interpretation)
+            command: The string or argument list to execute by /bin/sh (still with shell interpretation) FIXME?!
```

Comment:
FIXME?!

Reply:
I think I didn't like the string or argument list and wanted to change that, but it might be a bit too late for that :)